node/bindnode/node.go

```diff
@@ -0,0 +1,1099 @@
+package bindnode
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	ipld "github.com/ipld/go-ipld-prime"
+	basicnode "github.com/ipld/go-ipld-prime/node/basic"
+	"github.com/ipld/go-ipld-prime/schema"
+)
+
+// WrapNoSchema implements an ipld.Node given a pointer to a Go value.
+//
+// Same rules as PrototypeNoSchema apply.
+func WrapNoSchema(ptr interface{}) ipld.Node {
+	panic("TODO")
+	// ptrVal := reflect.ValueOf(ptr)
+	// if ptrVal.Kind() != reflect.Ptr {
+	// 	panic("must be a pointer")
+	// }
+	// return &_node{val: ptrVal.Elem()}
+}
+
+// Unwrap takes an ipld.Node implemented by one of the Wrap* APIs, and returns
+// a pointer to the inner Go value.
+//
+// Unwrap returns the input node if the node isn't implemented by this package.
+func Unwrap(node ipld.Node) (ptr interface{}) {
+	var val reflect.Value
+	switch node := node.(type) {
+	case *_node:
+		val = node.val
+	case *_nodeRepr:
+		val = node.val
+	default:
+		return node
+	}
+	if val.Kind() == reflect.Ptr {
+		panic("didn't expect val to be a pointer")
+	}
+	if !val.CanAddr() {
+		// Not addressable? Just return the interface as-is.
+		// TODO: This happens in some tests, figure out why.
+		return val.Interface()
+	}
+	return val.Addr().Interface()
+}
+
+// PrototypeNoSchema implements an ipld.NodePrototype given a Go pointer type.
+//
+// In this form, no IPLD schema is used; it is entirely inferred from the Go
+// type.
+//
+// Go types map to schema types in simple ways: Go string to schema String, Go
+// []byte to schema Bytes, Go struct to schema Map, and so on.
+//
+// A Go struct field is optional when its type is a pointer. Nullable fields are
+// not supported in this mode.
+func PrototypeNoSchema(ptrType interface{}) ipld.NodePrototype {
+	panic("TODO")
+	// typ := reflect.TypeOf(ptrType)
+	// if typ.Kind() != reflect.Ptr {
+	// 	panic("must be a pointer")
+	// }
+	// return &_prototype{goType: typ.Elem()}
+}
+
+// PrototypeOnlySchema implements an ipld.NodePrototype given an IPLD schema type.
+//
+// In this form, Go values are constructed with types inferred from the IPLD
+// schema, like a reverse of PrototypeNoSchema.
+func PrototypeOnlySchema(schemaType schema.Type) ipld.NodePrototype {
+	goType := inferGoType(schemaType)
+	return prototype(goType, schemaType)
+}
+
+// from IPLD Schema field names like "foo" to Go field names like "Foo".
+func fieldNameFromSchema(name string) string {
+	return strings.Title(name)
+}
+
+func inferGoType(typ schema.Type) reflect.Type {
```

Comment:
Short notes on what this is for would be helpful.  I see what it does, but harder to see _why_ it's necessary to do the things it does.

I think I see it mostly being used by... code that's related to unions?

Okay, and `PrototypeOnlySchema`.  Which I guess is pretty significant.

When I started writing this comment, I was wondering if the behaviour for unions is being used in a special enough way it should be separated from the rest of this, but I think I've talked myself back out of that already, so this is just a thinking-out-loud note now :)

Reply:
> Okay, and `PrototypeOnlySchema`. Which I guess is pretty significant.

Right, that's the only use case. When supplying both a schema and a Go type, inferGoType should never be used - that API entrypoint is just not properly done yet.

And yeah, unions are special because they need the extra "tag", but otherwise they are handled like any other type.